Improve depend from scratch

gasche opened the pull request:
This is a follow-up to the (temporarily paused) #8753 which makes it easier to run `alldepend` from a tree that has not been built yet. One commit fixes a bug in the `dynlink/Makefile` depend rule, the other removes an unnecessary dependency (`ocamlc`) to shorten the required build steps before alldepend can run correctly (the important dependencies are `coldstart` (or some subset of it), `ocamllex` and `ocamlyacc`).

Yesterday I committed incorrect `.depend` files that broke the parallel build. This was due to the `dynlink/Makefile` bug mentioned above.

gasche committed:
Makefile: ocamllex does not depend on ocamlc, it uses BOOT_OCAMLC

It seems that the dependency was introduced in 1998
  87b17301f4a408109a0f00f7a297b02c79724bc1
and it was already unnecessary at the time.

gasche committed:
fix dynlink/Makefile 'depend' target to work from non-built trees

Dynlink does a weird dance of copying compiler files from various
places. The 'depend' target does not perform this copying, and instead
just 'touch'es the files that should be present to run -- defined in
DEPEND_DUMMY_FILES: if you run 'depend' after a build, you get the
real content of the files, but if you run without building you get
empty files.

This is (dubious but apparently) correct for most of those files which
should not generate dependencies anyway, but not for
dynlink_platform_intf.mli which contains important dependencies (as
in: removing them breaks the parallel build) and should be properly
built before 'depend' runs.

gasche commented:
( cc @dra27 , who might still remember the context for this one )

dra27 commented on Makefile:
```diff
@@ -1021,7 +1021,7 @@ partialclean::
 
 .PHONY: ocamllex
-ocamllex: ocamlyacc ocamlc
+ocamllex: ocamlyacc
```
This was added in 87b173, but I don't think it was needed then, either.

dra27 reviewed (commented):
Oops - I think the bug in the dynlink `Makefile` was my fault. The changes look correct to me, although it shouldn't be necessary to build `ocamllex`. With these two further patches:

```diff
diff --git a/Makefile b/Makefile
index e005181f12..f935387d46 100644
--- a/Makefile
+++ b/Makefile
@@ -1074,6 +1074,9 @@ parsing/parser.ml: boot/menhir/parser.ml parsing/parser.mly \
 parsing/parser.mli: boot/menhir/parser.mli
        cat $< | sed "s/MenhirLib/CamlinternalMenhirLib/g" > $@
 
+beforedepend:: parsing/camlinternalMenhirLib.ml \
+  parsing/camlinternalMenhirLib.mli \
+       parsing/parser.ml parsing/parser.mli
 
 partialclean:: partialclean-menhir
 
diff --git a/ocamltest/Makefile b/ocamltest/Makefile
index e93c4ebb63..a4b67c58c1 100644
--- a/ocamltest/Makefile
+++ b/ocamltest/Makefile
@@ -184,7 +184,7 @@ ocamldep := $(ROOTDIR)/runtime/ocamlrun $(ROOTDIR)/boot/ocamlc -depend
 depflags := -slash
 depincludes :=
 
-ocamllex := $(ROOTDIR)/runtime/ocamlrun $(ROOTDIR)/lex/ocamllex
+ocamllex := $(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/boot/ocamllex
 
 ocamlyacc := $(ROOTDIR)/yacc/ocamlyacc
 
```

I'm getting no changes from `git clean -dfx ; ./configure --disable-debug-runtime --disable-instrumented-runtime ; make -j coldstart ocamlyacc ; make alldepend`

However, I don't know if @shindere has `ocamltest/Makefile` using the non-boot `ocamllex` for a reason.

gasche commented:
@dra27 your two extra patches look nice, would you commit them to my branch directly? (I like to preserve authorship but wouldn't want to write commit messages in your stead.)

dra27 committed:
Missing beforedepend dependencies for Menhir

Ensures that Menhir's sources are in place for a "from scratch"
`make alldepend`

dra27 committed:
Use boot ocamllex when building ocamltest

dra27 commented:
Sure! Please can @shindere sign off on the ocamltest `Makefile` change - I don't know if it using the built runtime and `ocamllex` for a reason (it may also have been because that part of the `Makefile` before all the many improvements and tidying to the build system)

shindere commented:
I can't remember why I didn't use `boot/ocamllex`, sorry.

I suspect I took inspiration from `ocamldoc` and also that, at that
time, it looked more sensible to me to use the just built `ocamllex` but
I am very fine with the change, thanks!

gasche commented:
I'm approving of @dra27's patches (I can't use the interface for this as I created the PR). This can be merged if someone approves of my own patches.

dra27 commented:
I think there are enough comments and approvals in all directions for all the various parts 🙂
